Add form notifications settings panel

vianasw opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Introduces a new notifications system that allows site administrators to specify which users should receive notifications when form responses are submitted.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
- Add `NotificationsSettings` component to manage notification recipients
- Implement `form-notifications` feature flag (default: disabled)
- Store notification recipients as a form-level attribute
- Add `notificationRecipients` attribute to Contact_Form defaults
- Update Feedback class to retrieve recipients from form attributes
- Add PHP filter `jetpack_forms_enable_notifications` to control the feature 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
### Prerequisites:
To enable this feature you have to add this filter:
```add_filter( 'jetpack_forms_enable_notifications', '__return_true' );```

* With the filter enabled, create a form and check that the sidebar has a section titled `Form notifications`.
<img width="304" height="642" alt="Screenshot 2025-10-08 at 10 44 23" src="https://github.com/user-attachments/assets/6ddb01bd-b8da-4b7a-9da7-b57dc303eef9" />

* Check that when enabled, the author of the post is set by default as the recipient of the notification.
<img width="297" height="788" alt="Screenshot 2025-10-08 at 10 44 34" src="https://github.com/user-attachments/assets/dd5ea87b-2378-44c1-8833-8331268b3fe7" />

* Add another user (like an editor) to the recipients list, save the draft, reload, and check that the recipients list has been persisted.
* Try to add to the recipients list a user who shouldn't have access to the post (like a subscriber), and confirm that it's not possible to do so.
* Disable the form response notifications and check that the recipients list collapses.
* Lastly, remove the feature flag filter and check that the `Form notifications` section is no longer present in the sidebar panel

vianasw committed:
Add form's notification settings panel

vianasw committed:
changelog

vianasw committed:
Add notification recipients to Feedback class

dhasilva committed:
remove boolean attribute and auto-select current author

vianasw committed:
Allow multiple users for form notifications

vianasw committed:
Add js unit tests

vianasw committed:
Add feature flag for form notifications

I've added the feature flag ``jetpack_forms_enable_notifications` to
enable/disable the whole form notification settings panel in the editor`

enejb commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-form-notifications-settings.js:
```diff
@@ -0,0 +1,107 @@
+import { __ } from '@wordpress/i18n';
+import InspectorHint from '../../shared/components/inspector-hint';
+
+const JetpackFormNotificationsSettings = ( { setAttributes, notificationRecipients } ) => {
```
Can we make this less verbouse. We are already in the JetpackForm contexts here.

enejb commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-form-notifications-settings.js:
```diff
@@ -0,0 +1,107 @@
+	return (
+		<>
+			<ToggleControl
+				label={ __( 'Enable form submission notifications', 'jetpack-forms' ) }
```
Instead of form submissions can we use form responses for more consistency across UI?

enejb commented on projects/packages/forms/src/blocks/contact-form/components/jetpack-form-notifications-settings.js:
```diff
@@ -0,0 +1,107 @@
+			{ localFormNotifications && (
+				<>
+					<InspectorHint>
+						{ __( 'Select users who can receive form submission notifications:', 'jetpack-forms' ) }
```
here too

enejb commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -166,6 +173,8 @@ private function load_from_post( WP_Post $feedback_post ) {
 		$this->ip_address = $parsed_content['ip'] ?? $this->get_first_field_of_type( 'ip' );
 		$this->subject    = $parsed_content['subject'] ?? $this->get_first_field_of_type( 'subject' );
 
+		$this->notification_recipients = $parsed_content['notification_recipients'] ?? $this->get_first_field_of_type( 'notification_recipients' );
```
I don't think we should store notification_recipients in a field. 

The reason why we this is the case for the `$this->get_first_field_of_type( 'subject' );` is because of legacy reasons. Sorry I should have made this more clear. 

Since only new forms would have this. We should be able to keep the data separated from the field data.

enejb commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -833,8 +863,9 @@ public function serialize() {
-				'ip'      => $this->ip_address,
+				'subject'                 => $this->subject,
+				'ip'                      => $this->ip_address,
+				'notification_recipients' => $this->notification_recipients,
```
This looks correct to me :)

enejb commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -1429,6 +1460,21 @@ private function get_computed_consent( $post_data, $form ) {
+	 * @return array
+	 */
+	private function get_computed_notification_recipients( $post_data, $form ) {
+		$field_ids = $form->get_field_ids();
```
🤔 Should this not be `$form->get_attribute( 'notification_recipients' ); ` ?

vianasw replied:
Yep. On it.

enejb commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -1429,6 +1460,21 @@ private function get_computed_consent( $post_data, $form ) {
+	private function get_computed_notification_recipients( $post_data, $form ) {
+		$field_ids = $form->get_field_ids();
+		if ( isset( $field_ids['notification_recipients'] ) ) {
+			return $this->get_field_value( $field_ids['notification_recipients'], $post_data );
```
$post_data here is the form submission so $_POST data. 
I don't think we should pass this info via from data. 

It should come via the JWT token which is searlized and unsearlized on form submission then available via the form attributes.

vianasw replied:
got it.

vianasw committed:
Rename JetpackFormNotificationsSettings

Make it less verbose since we're already in the JetpackForms context.
Renamed to NotificationsSettings.

vianasw committed:
Copy change: use response instead of submission

vianasw committed:
Make notification_recipients a form attribute

We don't need to store notification_recipients in a field like we're
doing with subject—which is done that way for legacy reasons. It should
come via the JWT token which is searlized and unsearlized on form
submission then available via the form attributes

vianasw committed:
Fix unit tests

vianasw committed:
Add unit tests for notifications handling

lezama commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -1429,6 +1460,26 @@ private function get_computed_consent( $post_data, $form ) {
+		if ( ! is_array( $notification_recipients ) ) {
+			return array();
+		}
+
```
Where do we perform server-side checks for user capabilities?

vianasw replied:
good point. Here is probably a good place.

vianasw replied:
I've added the server-side validation and some unit tests here: 20d10a52ee32403c971448b63d6379593013d650

lezama replied:
cool

vianasw committed:
Add server side notifications validation

Added server side validation of the notification recipients to make sure
they should receive the notifications—protection against frontend
tampering. It also covers the case where the user role might have
changed. I've added unit tests covering these cases

enejb commented on projects/packages/forms/src/contact-form/class-feedback.php:
```diff
@@ -688,6 +711,25 @@ public function get_subject() {
+	 * @param array $recipients The notification recipients.
+	 * @return void
+	 */
+	public function set_notification_recipients( $recipients ) {
```
As far as I can tell we don't use this anywhere.

vianasw committed:
Remove unused method

vianasw committed:
Remove unnecessary notification validation

We don't need to validate on load_from_post and we'll validate just
before sending the notifications

vianasw committed:
Remove test that doesn't make sense anymore
